crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1416,10 +1420,23 @@ impl<'db> ClassType<'db> {
         let DefinitionKind::Class(class_stmt_node) = definition.kind(db) else {
             panic!("Class type definition must have DefinitionKind::Class");
         };
-        class_stmt_node
-            .bases()
-            .iter()
-            .map(move |base_expr| definition_expression_ty(db, definition, base_expr))
+
+        // if there's a type params scope
+        let bases_scope_info: Option<(ScopeId<'db>, &TypeInference<'db>)> = self
+            .bases_specialized_scope(db)
+            .map(|bases_scope| (bases_scope, infer_scope_types(db, bases_scope)));
+
+        let mapper = move |base_expr: &ast::Expr| {
+            if let Some((bases_scope, inferences)) = bases_scope_info {
+                // when we have a specialized scope, we'll look up the inference
+                // within that scope
+                inferences.expression_ty(base_expr.scoped_ast_id(db, bases_scope))
+            } else {
+                // Otherwise, we can do the lookup based on the definition scope
+                definition_expression_ty(db, definition, base_expr)
+            }
+        };
+        class_stmt_node.bases().iter().map(mapper)
```

Comment:
Nit
```suggestion
        class_stmt_node.bases().iter().map(move |base_expr: &ast::Expr| {
            if let Some((bases_scope, inferences)) = bases_scope_info {
                // when we have a specialized scope, we'll look up the inference
                // within that scope
                inferences.expression_ty(base_expr.scoped_ast_id(db, bases_scope))
            } else {
                // Otherwise, we can do the lookup based on the definition scope
                definition_expression_ty(db, definition, base_expr)
            }
        })
```